storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3108,10 +3108,1655 @@ buf_block_t *fsp_get_latched_xdes_page(
     BUF_GET_POSSIBLY_FREED, mtr, err);
 }
 
+class spaceDefragmenter;
+
+/** Validate the file list node for the system tablespace.
+@param addr file space address
+@return true if validation successful or false */
+bool flst_node_offset_validate(const fil_addr_t *addr) noexcept
+{
+  return addr->boffset >= FIL_PAGE_DATA &&
+         addr->boffset < (srv_page_size - FIL_PAGE_DATA_END);
+}
+
+/** Prepare the steps for removing the file list node
+@param descr      descriptor to be removed
+@param free_limit maximum free limit in the tablespace
+@param prev_block previous block in the list
+@param next_block next block in the list
+@param mtr        mini-transaction
+@return error code */
+dberr_t flst_remove_prepare(xdes_t *descr, uint32_t free_limit,
+                            buf_block_t **prev_block,
+                            buf_block_t **next_block, mtr_t *mtr) noexcept
+{
+  fil_addr_t next_addr= flst_get_next_addr(descr + XDES_FLST_NODE);
+  fil_addr_t prev_addr= flst_get_prev_addr(descr + XDES_FLST_NODE);
+  dberr_t err= DB_SUCCESS;
+
+  if (next_addr.page != FIL_NULL)
+  {
+    *next_block= fsp_get_latched_page(page_id_t{0, next_addr.page}, mtr, &err);
+    if (err) return err;
```

Comment:
It should be more efficient to check something that must be in the return value register:
```c++
    *next_block= fsp_get_latched_page(page_id_t{0, next_addr.page}, mtr, &err);
    ut_ad(!*next_block == (err != DB_SUCCESS));
    if (!*next_block)
      return err;
```
The same pattern needs to be applied another time in this function and maybe as well in some other functions.